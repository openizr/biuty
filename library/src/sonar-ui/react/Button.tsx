/**
 * Copyright (c) Matthieu Jabbour. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import * as React from 'react';
import PropTypes, { InferProps } from 'prop-types';
import buildClass from 'sonar-ui/helpers/buildClass';

const propTypes = {
  id: PropTypes.string,
  icon: PropTypes.string,
  label: PropTypes.string,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
  modifiers: PropTypes.string,
  type: PropTypes.oneOf(['button', 'submit']),
  iconPosition: PropTypes.oneOf(['left', 'right']),
};

const defaultProps = {
  id: null,
  icon: null,
  label: null,
  onFocus: null,
  modifiers: '',
  type: 'button',
  onClick: undefined,
  iconPosition: 'left',
};

/**
 * Button.
 */
export default function UIButton(props: InferProps<typeof propTypes>): JSX.Element {
  // eslint-disable-next-line object-curly-newline
  const { label, icon, iconPosition, onClick, id, modifiers, type, onFocus } = props;
  const className = buildClass('ui-button', (modifiers as string).split(' '));
  const children = [
    (icon !== null) ? <i key="icon" className="ui-button__icon">{icon}</i> : null,
    (label !== null) ? <span key="label" className="ui-button__label">{label}</span> : null,
  ];

  const focusField = (): void => {
    if (onFocus !== undefined && onFocus !== null) {
      onFocus();
    }
  };

  return (
    <button
      id={id as string}
      onFocus={focusField}
      onClick={onClick as undefined}
      type={(type === 'submit') ? 'submit' : 'button'}
      className={`${className}${(icon !== null && label === null) ? ' ui-button--icon' : ''}`}
      tabIndex={((modifiers as string).includes('disabled') ? -1 : 0)}
    >
      {(iconPosition === 'left') ? children : children.reverse()}
    </button>
  );
}

UIButton.propTypes = propTypes;
UIButton.defaultProps = defaultProps;
UIButton.displayName = 'UIButton';
