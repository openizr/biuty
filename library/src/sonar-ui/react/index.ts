/**
 * Copyright (c) Matthieu Jabbour. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import UIP from 'sonar-ui/react/P';
import md from 'sonar-ui/helpers/md';
import UILink from 'sonar-ui/react/Link';
import UITitle from 'sonar-ui/react/Title';
import UIImage from 'sonar-ui/react/Image';
import UIButton from 'sonar-ui/react/Button';
import UISubtitle from 'sonar-ui/react/Subtitle';
import UIDropdown from 'sonar-ui/react/Dropdown';
import UITextarea from 'sonar-ui/react/Textarea';
import UICheckbox from 'sonar-ui/react/Checkbox';
import UITextfield from 'sonar-ui/react/Textfield';
import buildClass from 'sonar-ui/helpers/buildClass';
import UIFileUploader from 'sonar-ui/react/FileUploader';
import generateRandomId from 'sonar-ui/helpers/generateRandomId';
//
// eslint-disable-next-line import/prefer-default-export
export {
  UIP,
  UILink,
  UIButton,
  UIImage,
  UISubtitle,
  UIDropdown,
  UITitle,
  UITextfield,
  UITextarea,
  UICheckbox,
  UIFileUploader,
  md,
  buildClass,
  generateRandomId,
};
