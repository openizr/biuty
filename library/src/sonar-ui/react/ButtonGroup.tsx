/**
 * Copyright (c) Matthieu Jabbour. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import * as React from 'react';
import PropTypes, { InferProps } from 'prop-types';

const propTypes = {
  id: PropTypes.string,
  // icon: PropTypes.string,
  // label: PropTypes.string,
  // onClick: PropTypes.func,
  // modifiers: PropTypes.string,
  // iconPosition: PropTypes.oneOf(['left', 'right']),
  children: PropTypes.oneOfType([PropTypes.object, PropTypes.arrayOf(PropTypes.object)]),
};

const defaultProps = {
  id: null,
  // icon: null,
  // label: null,
  // onClick: undefined,
  // iconPosition: 'left',
  // modifiers: 'contained',
  children: null,
};

/**
 * Button group.
 */
export default function UIButtonGroup(props: InferProps<typeof propTypes>): JSX.Element {
  const { id, children } = props;
  return (
    <div
      id={id as string}
      className="ui-button-group"
    >
      {children}
    </div>
  );
}

UIButtonGroup.propTypes = propTypes;
UIButtonGroup.defaultProps = defaultProps;
UIButtonGroup.displayName = 'UIButtonGroup';
