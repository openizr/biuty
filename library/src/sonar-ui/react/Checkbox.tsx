/**
 * Copyright (c) KivFinance, Inc.
 * All rights reserved.
 */

import * as React from 'react';
import PropTypes, { InferProps } from 'prop-types';
import buildClass from 'sonar-ui/helpers/buildClass';
import generateRandomId from 'sonar-ui/helpers/generateRandomId';

const propTypes = {
  id: PropTypes.string,
  value: PropTypes.bool,
  label: PropTypes.string,
  helper: PropTypes.string,
  onChange: PropTypes.func,
  readonly: PropTypes.bool,
  modifiers: PropTypes.string,
  name: PropTypes.string.isRequired,
};

const defaultProps = {
  id: null,
  label: null,
  helper: null,
  value: false,
  modifiers: '',
  onChange: null,
  readonly: false,
};

/**
 * Checkbox.
 */
export default function UICheckbox(props: InferProps<typeof propTypes>): JSX.Element {
  // eslint-disable-next-line object-curly-newline
  const { id, modifiers, label, helper, onChange, value, name, readonly } = props;
  const [randomId] = React.useState(generateRandomId());
  const [currentValue, setCurrentValue] = React.useState(value);
  const className = buildClass('ui-checkbox', (modifiers as string).split(' '));

  // Updates current value each time the `value` property is changed.
  React.useEffect(() => {
    setCurrentValue(value);
  }, [value]);

  const changeValue = (event: React.ChangeEvent<HTMLInputElement>): void => {
    setCurrentValue(event.target.checked);
    if (onChange !== undefined && onChange !== null) {
      onChange(event.target.checked);
    }
  };

  return (
    <div
      id={id as string}
      className={className}
    >
      {(label !== null) ? <label className="ui-checkbox__label" htmlFor={randomId}>{label}</label> : null}
      <input
        name={name}
        id={randomId}
        type="checkbox"
        readOnly={readonly as boolean}
        checked={currentValue as boolean}
        onChange={(readonly === false) ? changeValue : undefined}
        tabIndex={((modifiers as string).includes('disabled') ? -1 : 0)}
      />
      {(helper !== null) ? <span className="ui-checkbox__helper">{helper}</span> : null}
    </div>
  );
}

UICheckbox.propTypes = propTypes;
UICheckbox.defaultProps = defaultProps;
UICheckbox.displayName = 'UICheckbox';
