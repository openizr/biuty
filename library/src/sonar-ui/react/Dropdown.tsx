/**
 * Copyright (c) Matthieu Jabbour. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import * as React from 'react';
import PropTypes, { InferProps } from 'prop-types';
import buildClass from 'sonar-ui/helpers/buildClass';
import generateRandomId from 'sonar-ui/helpers/generateRandomId';

const optionType = {
  type: PropTypes.oneOf(['header', 'divider', 'option']).isRequired,
  value: PropTypes.string,
  label: PropTypes.string,
  disabled: PropTypes.bool,
};

const propTypes = {
  id: PropTypes.string,
  multiple: PropTypes.bool,
  helper: PropTypes.string,
  options: PropTypes.arrayOf(PropTypes.shape(optionType).isRequired).isRequired,
  label: PropTypes.string,
  modifiers: PropTypes.string,
  clearIcon: PropTypes.string,
  defaultOptions: PropTypes.arrayOf(PropTypes.string.isRequired),
  onChange: PropTypes.func,
};

const defaultProps = {
  id: null,
  label: null,
  helper: null,
  multiple: false,
  clearIcon: null,
  defaultOptions: [],
  modifiers: 'contained',
  onChange: (): null => null,
};

type options = InferProps<typeof optionType>[];

const findSiblingOption = (
  options: options,
  currentIndex: number,
  direction: number,
  lastOption: (number | null) = null,
): number => {
  const nextIndex = currentIndex + direction;
  if (nextIndex < 0 || nextIndex >= options.length) {
    return lastOption || currentIndex;
  }
  const option = options[nextIndex];
  return (option.value !== undefined && option.disabled !== true)
    ? nextIndex
    : findSiblingOption(options, nextIndex, direction, lastOption || currentIndex);
};

/**
 * Dropdown.
 */
export default function UIDropdown(props: InferProps<typeof propTypes>): JSX.Element {
  // TODO WE DON'T WANT TO TRIGGER ONCHANGE WHEN DEFAULT OPTION IS UPDATED AND NO OPTION IS SELECTED
  // eslint-disable-next-line object-curly-newline
  const { options, id, modifiers, defaultOptions, multiple, label, helper, clearIcon,
    onChange,
  } = props;
  const [randomId] = React.useState(generateRandomId());
  const [position, setPosition] = React.useState('bottom');
  const [isDisplayed, setIsDisplayed] = React.useState(false);
  const [focusedOption, setFocusedOption] = React.useState(-1);
  const [mapping, setMapping] = React.useState({} as Record<string, string>);
  const [selectedOptions, setSelectedOptions] = React.useState(defaultOptions as string[]);
  const classes = buildClass('ui-dropdown', (modifiers as string).split(' '));
  const buttonRef = React.useRef<HTMLButtonElement>(null);
  const ulRef = React.useRef<HTMLUListElement>(null);
  const initial = React.useRef(false);

  // Avoids having to pass through the entire options array at each rendering.
  React.useEffect(() => {
    setMapping(options.reduce((finalMapping, option) => {
      if (option.value !== undefined) {
        return { ...finalMapping, [option.value as string]: option.label };
      }
      return finalMapping;
    }, {}));
  }, [options]);

  React.useEffect(() => {
    (onChange as any)(selectedOptions);
  }, [selectedOptions]);

  React.useEffect(() => {
    if (
      selectedOptions.length === 0
      && JSON.stringify(selectedOptions) !== JSON.stringify(defaultOptions)) {
      setSelectedOptions(defaultOptions as string[]);
    }
  }, [defaultOptions]);

  // HTML elements with `display: none` can't be focused. Thus, we need to wait for the HTML list to
  // be displayed before actually focusing it.
  React.useEffect(() => {
    if (ulRef.current !== null && isDisplayed === true) {
      const focusedIndex = (focusedOption >= 0) ? focusedOption : findSiblingOption(options, 0, +1);
      (ulRef.current.childNodes[focusedIndex] as HTMLElement).focus();
      setFocusedOption(focusedIndex);
    } else if (isDisplayed === false && buttonRef.current !== null && initial.current === true) {
      buttonRef.current.focus();
    }
  }, [isDisplayed]);

  React.useEffect(() => {
    if (ulRef.current !== null && focusedOption >= 0 && isDisplayed === true) {
      (ulRef.current.childNodes[focusedOption] as HTMLElement).focus();
    }
  }, [focusedOption]);

  React.useEffect(() => {
    initial.current = true;
  }, []);

  const clearOptions = (): void => {
    setSelectedOptions([options[findSiblingOption(options, 0, +1)].value as string]);
  };

  const displayList = (): void => {
    if (buttonRef.current !== null) {
      const relativeOffsetTop = buttonRef.current.getBoundingClientRect().top;
      setPosition((relativeOffsetTop > window.innerHeight / 2) ? 'top' : 'bottom');
      setIsDisplayed(true);
    }
  };

  const hideList = (force = false) => (event: React.FocusEvent<HTMLElement> | null): void => {
    // We first ensure that the newly focused element is not an option of the list.
    const focusIsOutsideList = (event !== null)
      ? !event.currentTarget.contains(event.relatedTarget as Node)
      : true;
    if (focusIsOutsideList && (force === true || multiple === false)) {
      setIsDisplayed(false);
    }
  };

  const toggleOption = (optionIndex: number) => (): void => {
    const selectedOption = options[optionIndex].value as string;
    if (multiple === false) {
      setSelectedOptions([selectedOption]);
    } else {
      const selectedIndex = selectedOptions.indexOf(selectedOption);
      setSelectedOptions((selectedIndex >= 0)
        ? selectedOptions.slice(0, selectedIndex).concat(selectedOptions.slice(selectedIndex + 1))
        : selectedOptions.concat([selectedOption]));
    }
  };

  const clickOnOption = (optionIndex: number) => (): void => {
    setFocusedOption(optionIndex);
    toggleOption(optionIndex)();
    hideList()(null);
  };

  const navigate = (event: React.KeyboardEvent<HTMLElement>): void => {
    if (ulRef.current !== null) {
      const { key } = event;
      switch (key) {
        case 'ArrowUp':
          setFocusedOption(findSiblingOption(options, focusedOption, -1));
          event.preventDefault();
          break;
        case 'ArrowDown':
          setFocusedOption(findSiblingOption(options, focusedOption, +1));
          event.preventDefault();
          break;
        case 'PageUp':
        case 'Home':
          setFocusedOption(findSiblingOption(options, 0, +1));
          event.preventDefault();
          break;
        case 'End':
        case 'PageDown':
          setFocusedOption(findSiblingOption(options, options.length, -1));
          event.preventDefault();
          break;
        case ' ':
        case 'Enter':
          if (isDisplayed === false) {
            setIsDisplayed(true);
          } else {
            toggleOption(focusedOption)();
            hideList()(null);
          }
          break;
        case 'Escape':
          hideList(true)(null);
          break;
        default:
          break;
      }
    }
  };

  const listModifiers = ['', position].concat((isDisplayed === true) ? ['expanded'] : []);
  return (
    <div
      id={id as string}
      className={classes}
    >
      {(label !== null)
        ? <label className="ui-dropdown__label" htmlFor={randomId}>{label}</label>
        : null}
      <div className="ui-dropdown__wrapper">

        <button
          type="button"
          id={randomId}
          ref={buttonRef}
          onKeyDown={navigate}
          aria-haspopup="listbox"
          onMouseDown={displayList}
          className="ui-dropdown__wrapper__field"
          aria-labelledby={`${randomId} ${randomId}`}
          tabIndex={((modifiers as string).includes('disabled') ? -1 : 0)}
        >
          {selectedOptions.map((value) => mapping[value]).join(', ')}
        </button>
        {(clearIcon !== null)
          ? (
            <button
              type="button"
              id={`${randomId}c`}
              onClick={clearOptions}
              className="ui-dropdown__wrapper__clear-button"
              aria-labelledby={`${randomId} ${randomId}c`}
              tabIndex={((modifiers as string).includes('disabled') ? -1 : 0)}
            >
              <i className="ui-dropdown__wrapper__clear-button__icon">{clearIcon}</i>
            </button>
          )
          : null}
        <ul
          ref={ulRef}
          tabIndex={-1}
          role="listbox"
          onKeyDown={navigate}
          onBlur={hideList(true)}
          aria-labelledby={randomId}
          aria-expanded={isDisplayed}
          aria-activedescendant={`${randomId}${focusedOption}`}
          className={buildClass('ui-dropdown__wrapper__list', listModifiers)}
        >
          {options.map((option, index) => {
            const key = randomId + index;
            // Divider...
            if (option.type === 'divider') {
              return <li key={key} className="ui-dropdown__wrapper__list__divider" />;
            }
            // Group's header...
            if (option.type === 'header') {
              return <li key={key} className="ui-dropdown__wrapper__list__header">{option.label}</li>;
            }
            // Disabled option...
            if (option.disabled === true) {
              return (
                <li
                  key={key}
                  className="ui-dropdown__wrapper__list__option ui-dropdown__wrapper__list__option--disabled"
                >
                  {option.label}
                </li>
              );
            }
            // Classic option...
            const optionModifiers = selectedOptions.includes(option.value as string) ? ['selected'] : [];
            return (
              <li
                id={key}
                key={key}
                role="option"
                tabIndex={-1}
                onMouseDown={clickOnOption(index)}
                aria-selected={focusedOption === index}
                className={buildClass('ui-dropdown__wrapper__list__option', optionModifiers)}
              >
                {option.label}
              </li>
            );
          })}
        </ul>
      </div>
      {(helper !== null) ? <p className="ui-dropdown__wrapper__helper">{helper}</p> : null}
    </div>
  );
}

UIDropdown.propTypes = propTypes;
UIDropdown.defaultProps = defaultProps;
UIDropdown.displayName = 'UIDropdown';
