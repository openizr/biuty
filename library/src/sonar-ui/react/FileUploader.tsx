/**
 * Copyright (c) Matthieu Jabbour. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import * as React from 'react';
import PropTypes, { InferProps } from 'prop-types';
import buildClass from 'sonar-ui/helpers/buildClass';
import generateRandomId from 'sonar-ui/helpers/generateRandomId';

const propTypes = {
  id: PropTypes.string,
  icon: PropTypes.string,
  label: PropTypes.string,
  onFocus: PropTypes.func,
  helper: PropTypes.string,
  onChange: PropTypes.func,
  multiple: PropTypes.bool,
  modifiers: PropTypes.string,
  placeholder: PropTypes.string,
  name: PropTypes.string.isRequired,
  iconPosition: PropTypes.oneOf(['left', 'right']),
};

const defaultProps = {
  id: null,
  icon: null,
  label: null,
  helper: null,
  onFocus: null,
  modifiers: '',
  multiple: false,
  placeholder: null,
  iconPosition: 'left',
  onChange: (): null => null,
};

/**
 * File uploader.
 */
export default function UIFileUploader(props: InferProps<typeof propTypes>): JSX.Element {
  const {
    id, modifiers, label, helper, iconPosition, icon,
    onChange, multiple, name, placeholder, onFocus,
  } = props;
  const [randomId] = React.useState(generateRandomId);
  const [currentValue, setCurrentValue] = React.useState([] as string[]);
  const className = buildClass('ui-file-uploader', (modifiers as string).split(' '));

  const changeValue = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const files = [];
    const numberOfFiles = (event.target.files?.length || 0);
    for (let index = 0; index < numberOfFiles; index += 1) {
      files.push((event.target.files as FileList)[index]);
    }
    setCurrentValue(files.map((file) => file.name));
    if (onChange !== undefined && onChange !== null) {
      onChange(files);
    }
  };

  const focusField = (): void => {
    if (onFocus !== undefined && onFocus !== null) {
      onFocus();
    }
  };

  const children = [
    (icon !== null) ? <i key="icon" className="ui-button__icon">{icon}</i> : null,
    <input
      type="file"
      name={name}
      id={randomId}
      key={randomId}
      onFocus={focusField}
      onChange={changeValue}
      multiple={multiple as boolean}
      className="ui-file-uploader__field"
    />,
  ];
  return (
    <div
      id={id as string}
      className={className}
    >
      {(label !== null)
        ? <label className="ui-file-uploader__label" htmlFor={randomId}>{label}</label>
        : null}
      {(iconPosition === 'left') ? children : children.reverse()}
      <span className="ui-file-uploader__files-list">{(currentValue.length === 0) ? placeholder : currentValue.join(', ')}</span>
      {(helper !== null) ? <p className="ui-file-uploader__helper">{helper}</p> : null}
    </div>
  );
}

UIFileUploader.propTypes = propTypes;
UIFileUploader.defaultProps = defaultProps;
UIFileUploader.displayName = 'UIFileUploader';
