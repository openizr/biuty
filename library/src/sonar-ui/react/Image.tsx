/**
 * Copyright (c) Matthieu Jabbour. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import * as React from 'react';
import buildClass from 'sonar-ui/helpers/buildClass';
import PropTypes, { InferProps } from 'prop-types';

const propTypes = {
  src: PropTypes.string.isRequired,
  alt: PropTypes.string.isRequired,
  ratio: PropTypes.string.isRequired,
  modifiers: PropTypes.string,
  itemProp: PropTypes.string,
};

const defaultProps = {
  modifiers: '',
  itemProp: null,
};

const getDimensions = (ratio: string): { width: number; height: number; } => {
  let dimensions;
  switch (ratio) {
    case 'square':
      return { width: 1, height: 1 };
    case 'portrait':
      return { width: 2, height: 3 };
    case 'landscape':
      return { width: 3, height: 2 };
    case 'panoramic':
      return { width: 16, height: 9 };
    default:
      dimensions = ratio.split(':').map((value) => parseInt(value, 10));
      return { width: dimensions[0], height: dimensions[1] };
  }
};

/**
 * Image.
 */
export default function UIImage(props: InferProps<typeof propTypes>): JSX.Element {
  // eslint-disable-next-line object-curly-newline
  const { src, alt, ratio, modifiers, itemProp } = props;
  const dimensions = getDimensions(ratio);
  const className = buildClass('ui-image', `${ratio.replace(':', 'x')} ${modifiers}`.split(' '));

  // Custom aspect ratio...
  if (/^([0-9]+):([0-9]+)$/i.test(ratio) === true) {
    return (
      <img
        src={src}
        alt={alt}
        loading="lazy"
        className={className}
        width={dimensions.width}
        height={dimensions.height}
        itemProp={(itemProp !== null) ? itemProp : undefined}
      />
    );
  }

  // Standard aspect ratio...
  return (
    <div className={className}>
      <img
        src={src}
        alt={alt}
        loading="lazy"
        width={dimensions.width}
        height={dimensions.height}
        itemProp={(itemProp !== null) ? itemProp : undefined}
      />
    </div>
  );
}

UIImage.displayName = 'UIImage';
UIImage.propTypes = propTypes;
UIImage.defaultProps = defaultProps;
