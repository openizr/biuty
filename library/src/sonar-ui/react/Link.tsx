/**
 * Copyright (c) Matthieu Jabbour. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import * as React from 'react';
import md from 'sonar-ui/helpers/md';
import PropTypes, { InferProps } from 'prop-types';
import buildClass from 'sonar-ui/helpers/buildClass';

const propTypes = {
  id: PropTypes.string,
  onClick: PropTypes.func,
  modifiers: PropTypes.string,
  href: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
};

const defaultProps = {
  id: null,
  modifiers: '',
  onClick: (): null => null,
};

/**
 * Link.
 */
export default function UILink(props: InferProps<typeof propTypes>): JSX.Element {
  // eslint-disable-next-line object-curly-newline
  const { id, href, label, onClick, modifiers } = props;

  return (
    // eslint-disable-next-line jsx-a11y/control-has-associated-label
    <a
      id={id as string}
      href={href}
      onClick={onClick as () => null}
      className={buildClass('ui-link', (modifiers as string).split(' '))}
      // eslint-disable-next-line react/no-danger
      dangerouslySetInnerHTML={{ __html: md(label) }}
    />
  );
}

UILink.propTypes = propTypes;
UILink.defaultProps = defaultProps;
UILink.displayName = 'UILink';
