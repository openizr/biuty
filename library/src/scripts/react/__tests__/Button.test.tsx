/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @jest-environment jsdom
 */

import React from 'react';
import UIButton from 'scripts/react/Button';
import { render } from '@testing-library/react';

const JSXUIButton = UIButton as JSXElement;

describe('react/UIButton', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders correctly - basic', () => {
    const { container } = render(<JSXUIButton label="Test" modifiers="large" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - with id', () => {
    const { container } = render(<JSXUIButton label="Test" id="test" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - no label', () => {
    const { container } = render(<JSXUIButton />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - label with icon', () => {
    const { container } = render(<JSXUIButton label="Test" icon="star" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - default type', () => {
    const { container } = render(<JSXUIButton type={null} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - type submit', () => {
    const { container } = render(<JSXUIButton type="submit" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - icon only', () => {
    const { container } = render(<JSXUIButton icon="star" iconPosition={null} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - icon only', () => {
    const { container } = render(<JSXUIButton icon="star" iconPosition="right" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - disabled', () => {
    const { container } = render(<JSXUIButton modifiers="disabled" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
