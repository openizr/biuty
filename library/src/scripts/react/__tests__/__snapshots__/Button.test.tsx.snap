// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react/UIButton renders correctly - basic 1`] = `
<div>
  <button
    class="ui-button ui-button--large"
    tabindex="0"
    type="button"
  >
    <span
      class="ui-button__label"
    >
      Test
    </span>
  </button>
</div>
`;

exports[`react/UIButton renders correctly - disabled 1`] = `
<div>
  <button
    class="ui-button ui-button--disabled"
    tabindex="-1"
    type="button"
  />
</div>
`;

exports[`react/UIButton renders correctly - icon only 1`] = `
<div>
  <button
    class="ui-button ui-button--icon"
    tabindex="0"
    type="button"
  >
    <i
      class="ui-button__icon"
    >
      star
    </i>
  </button>
</div>
`;

exports[`react/UIButton renders correctly - label with icon 1`] = `
<div>
  <button
    class="ui-button"
    tabindex="0"
    type="button"
  >
    <i
      class="ui-button__icon"
    >
      star
    </i>
    <span
      class="ui-button__label"
    >
      Test
    </span>
  </button>
</div>
`;

exports[`react/UIButton renders correctly - no label 1`] = `
<div>
  <button
    class="ui-button"
    tabindex="0"
    type="button"
  />
</div>
`;

exports[`react/UIButton renders correctly - right icon 1`] = `
<div>
  <button
    class="ui-button ui-button--icon"
    tabindex="0"
    type="button"
  >
    <i
      class="ui-button__icon"
    >
      star
    </i>
  </button>
</div>
`;

exports[`react/UIButton renders correctly - type submit 1`] = `
<div>
  <button
    class="ui-button"
    tabindex="0"
    type="submit"
  >
    <span
      class="ui-button__label"
    >
      Test
    </span>
  </button>
</div>
`;

exports[`react/UIButton renders correctly - with id and listeners 1`] = `
<div>
  <button
    class="ui-button ui-button--large"
    id="test"
    tabindex="0"
    type="button"
  >
    <span
      class="ui-button__label"
    >
      Test
    </span>
  </button>
</div>
`;
